core/src/server/rpc_module.rs

```diff
@@ -777,47 +788,46 @@ impl SubscriptionSink {
 		S: Stream<Item = T> + Unpin,
 		T: Serialize,
 	{
-		let mut close_stream = self.close.clone();
-		let mut item = stream.next();
-		let mut close = close_stream.next();
-
-		loop {
-			match futures_util::future::select(item, close).await {
-				Either::Left((Some(result), c)) => {
-					match self.send(&result) {
-						Ok(_) => (),
-						Err(Error::SubscriptionClosed(close_reason)) => {
-							self.close(&close_reason);
-							break Ok(());
-						}
-						Err(err) => {
-							tracing::error!("subscription `{}` failed to send item got error: {:?}", self.method, err);
-							break Err(err);
-						}
-					};
-					close = c;
-					item = stream.next();
-				}
-				// No messages should be sent over this channel
-				// if that occurred just ignore and continue.
-				Either::Right((Some(_), i)) => {
-					item = i;
-					close = close_stream.next();
-				}
-				// Connection terminated.
-				Either::Right((None, _)) => {
-					self.close(&SubscriptionClosed::new(SubscriptionClosedReason::ConnectionReset));
-					break Ok(());
+		if let Some(close_notify) = self.close_notify.clone() {
```

Comment:
why do we need the clone here @dvdplm?!

Reply:
I can't move out the `Arc<Notify>` and I can't use a reference either (because `send` takes a mutable reference below). :/

Reply:
should be possible with `take` here but it didn't work when I tried so I just wonder why :)

Reply:
`take()` compiles but then the tests fail. :/

Reply:
Played a bit with this, and not sure there is a super clean solution. You either clone it here, or somehow decouple the `closed_fut` from `self` lifetime wise, which is really hard with pinned futures.